beacon-chain/attestation/service.go

```diff
@@ -140,20 +140,42 @@ func (a *Service) attestationPool() {
 			return
 		// Listen for a newly received incoming attestation from the sync service.
 		case attestation := <-a.incomingChan:
-			enc, err := proto.Marshal(attestation)
-			if err != nil {
-				log.Errorf("Could not marshal incoming attestation to bytes: %v", err)
-				continue
-			}
-			h := hashutil.Hash(enc)
+			safelyHandleAttestation(a.handleAttestation, attestation)
+		}
+	}
+}
 
-			if err := a.updateLatestAttestation(a.ctx, attestation); err != nil {
-				log.Errorf("Could not update attestation pool: %v", err)
-				continue
+// safelyHandleMessage will recover and log any panic that occurs from the
+// function argument.
+func safelyHandleAttestation(fn func(att *pb.Attestation), att *pb.Attestation) {
```

Comment:
Is there any way we can abstract this handling logic to be reused? 

The one that we have for regular sync is a generic proto.Message, maybe it can be moved to shared and reused?

Reply:
we cant change the ones in regular sync and initial sync since the message types are `p2p.Message` and abstracting the functions into a `proto.Message` function does not make sense as the sync services methods need the `peer` and `ctx` for syncing and tracing respectively.

 I have however been able to abstract for the other services into a common function